api_docs: Add roles and permissions page to API documentation.

laurynmm opened the pull request:
Adds a non-endpoint specific page to the API documentation about organization-level roles and permissions for users in order to highlight important and useful information for clients and API users.

See [this CZO topic](https://chat.zulip.org/#narrow/stream/412-api-documentation/topic/stream_post_policy.20.28API.20docs.29/near/1323762) for more context.

**Screenshot of new API Roles and permissions documentation page**:
![Screenshot from 2022-03-30 16-16-57](https://user-images.githubusercontent.com/63245456/160856956-6cc232dd-9c63-4dd4-bda1-d1b27c362950.png)

gnprice commented on templates/zerver/api/roles-and-permissions.md:
```diff
@@ -0,0 +1,85 @@
+# Roles and permissions
+
+Zulip offers [several levels of permissions based on user
+organization-level roles](/help/roles-and-permissions).
```
Probably easier to parse if "user organization-level roles" is unpacked: e.g. "a user's role in a Zulip organization".

timabbott replied:
I'd also shrink the link to just be on "user's role", not the whole sentence.

gnprice commented on templates/zerver/api/roles-and-permissions.md:
```diff
@@ -0,0 +1,85 @@
+Here are some important details to note when working with these
+roles and permissions in Zulip's API:
+
+## A user's organization-level role
```
This version works fine, though. I think in "user organization-level roles", I get garden-pathed into wanting "user organization" to be a phrase.

laurynmm replied:
Actually, I think shortening to 'A user's role' would work as I work through these revisions.

There's one other part of the API that uses the word 'role' as a property / in regards to streams (is the user a subscriber or an admin of the stream). But if we're clear in the documentation for those instances and have this page to link to for clarification, I think it will be fine.

gnprice commented on templates/zerver/api/roles-and-permissions.md:
```diff
@@ -0,0 +1,85 @@
+## A user's organization-level role
+
+As of [Zulip 4.0](/api/changelog#changes-in-zulip-40) (feature levels
+59 and 60), endpoints and [events](/api/get-events#realm_user-add)
+that return user account data include a `role` property, which
+contains the user's organization-level role. These roles are encoded as:
+
+* Organization owner: 100
+
+* Organization administrator: 200
+
+* Organization moderator: 300
+
+* Member: 400
+
+* Guest: 600
+
+User account data also include these relevant boolean properties:
+
+* `is_admin` specifying whether the user is an organization administrator.
+
+* `is_guest` specifying whether the user is a guest user.
+
+* `is_owner` specifying whether the user is an organization owner, and
+  present as of [Zulip 3.0](/api/changelog#changes-in-zulip-30) (feature
+  levels 8 and 11). If true, `is_admin` will also be true.
+
+* `is_billing_admin` specifying whether the user is a billing administrator,
+  and present as of [Zulip 5.0](/api/changelog#changes-in-zulip-50)
+  (feature level 73).
+
+Additionally, as of [Zulip 4.0](/api/changelog#changes-in-zulip-40)
+(feature level 60), when `realm_user` is present in `fetch_event_types`,
+the [`/register`endpoint](/api/register-queue) will return an
+`is_moderator` boolean value specifying whether the user is an organization
+moderator.
+
+Note that the [`GET users/me`](/api/get-own-user) endpoint did not
+return a `is_guest` value until
+[Zulip 3.0](/api/changelog#changes-in-zulip-30) (feature level 10).
```
All of these details should also exist on the respective endpoint docs. I think it's probably best to leave the exact details to be found there, and have this page just mention these properties and then mention the different places they're found, with links to get the details (`/api/get-events#realm_user-add`, `/api/get-own-user`, and so on.)

An alternative would be to take the details out of those pages and centralize them here, and have those link here. But that gets sticky because the details don't always match between different endpoints. Also because there are like 6 different properties, and so for a given endpoint to have each of those link to this central doc wouldn't be hugely shorter or simpler than just including the details directly.

gnprice replied:
Also when mentioning the different places these properties are found, should specifically call out that they appear on the /api/register-queue response to describe the self-user.

timabbott replied:
Yeah, I agree with avoiding the duplication here, and that a list of endpoints that reference roles would be a good replacement. It clutters the narrative to have it here, and wherever we duplicate information, we create opportunities to have mistakes.

(And that the individual endpoints are the right place to have the single copy of the feature level cutoffs, well, the copy that isn't the API changelog itself, which tries to be brief and just reference the per-endpoint documentation as well)

gnprice commented on templates/zerver/api/roles-and-permissions.md:
```diff
@@ -0,0 +1,85 @@
+was created, to determine if a user has the permissions of a full member
+or a new member.
+
+Note that the [`GET users/me`](/api/get-own-user) endpoint did not
+return a `date_joined` value until
+[Zulip 3.0](/api/changelog#changes-in-zulip-30) (feature level 10).
```
I think this can also be left unmentioned here, and something to find on the endpoint doc when it's relevant.

(For a full-blown client, it won't be relevant because they're getting the user data from the event system with /api/register-queue, not from separate fetches like /api/get-own-user. For a simple bot, they're unlikely to worry about supporting ancient versions.)

timabbott replied:
Yeah, I agree with this.

gnprice reviewed (commented):
Thanks @laurynmm! This will be very helpful. Comments below.

Also we'll want to add links to this page. I see this is WIP, so I figure you're already planning to do that in a future draft. I think the first places I'd want it are (a) anywhere we have an existing link to the corresponding user-facing help page /help/roles-and-permissions, and (b) anywhere we have a "something_policy" property with values corresponding to these roles, especially any that have "full members" as an option because that one is particularly nonobvious how to compute before one finds this page.

timabbott commented on templates/zerver/api/roles-and-permissions.md:
```diff
@@ -0,0 +1,85 @@
+
+* Guest: 600
+
+User account data also include these relevant boolean properties:
```
I think it's worth explicitly mentioning that all of these except for `is_billing_admin` duplicate of the `role` field, present for help readability in simpler clients.  (Which might be best done by talking about `is_billing_admin` separately).

timabbott commented on templates/zerver/api/roles-and-permissions.md:
```diff
@@ -0,0 +1,85 @@
+
+The most trusted, and therefore most restrictive, permission level or policy
+would be organization owner only, or nobody. And the least trusted, and
+therefore least restrictive, permission level would be members only,
```
This makes it sound like "members only" == "everyone", which is not true. It might be simpler to actually just list the possiblities:
* Everyone (least restrictive)
* Members and above
* Full members and above 
* ...
* Nobody (most restrictive)

laurynmm commented:
@timabbott and @gnprice - I appreciate the thorough feedback for that first draft. I've updated the new documentation page for your review comments. And added / updated the links in the OpenAPI documentation for this new page. Note that there were a few places in the OpenAPI documentation where it made sense to keep the link to the help center page on roles and permissions.

Here's [the HTML diff](https://pastebin.com/Yv7Cn8wd) for the API docs, with the redundant left sidebar diffs removed. This shows the updated links, which pages are impacted, and confirms that the reference links are all rendering correctly.

In the **A user's role** section, I added a sub-header for listing the relevant endpoints: **User account data in the API**. I tried a couple of different page structures / orders and landed on this one, but there might be an option I didn't think of or that you were envisioning. Without the sub-header, the text for the roles section gets a bit long. That being said, I'm not sure I landed on the best text / title for that sub-header / section.

**Screenshot of updated API Roles and permissions page**:
![Screenshot from 2022-04-04 13-41-40](https://user-images.githubusercontent.com/63245456/161536632-f828188b-e774-4332-9cf9-0cecbf7b58f8.png)

timabbott commented:
I merged this with a few small changes. Huge thanks for doing this @laurynmm!

```
index 67a59fff58..1fd005f772 100644
--- a/templates/zerver/api/roles-and-permissions.md
+++ b/templates/zerver/api/roles-and-permissions.md
@@ -30,16 +30,19 @@ the related roles above:
 
 * `is_guest` specifying whether the user is a guest user.
 
+These are intended as conveniences for simple clients, and clients
+should prefer using the `role` field, since only that one is updated
+by the [events API](/api/get-events).
 
-Note that the [`POST /register`endpoint](/api/register-queue) also
-returns an `is_moderator` boolean property specifying whether the
-current user is an organization moderator.
+Note that [`POST /register`](/api/register-queue) also returns an
+`is_moderator` boolean property specifying whether the current user is
+an organization moderator.
 
 Additionally, user account data include an `is_billing_admin` property
 specifying whether the user is a billing administrator for the Zulip
 organization, which is not related to one of the roles listed above,
 but rather allows for specific permissions related to billing
-administration in [paid Zulip plans](https://zulip.com/plans/).
+administration in [paid Zulip Cloud plans](https://zulip.com/plans/).
 
 ### User account data in the API
 
@@ -76,7 +79,7 @@ deletion](/help/configure-message-editing-and-deletion) and
 [streams permissions](/help/stream-permissions). The potential
 permission levels are:
 
-* Everyone / Any user (least restrictive)
+* Everyone / Any user including Guests (least restrictive)
 
 * Members
 
```
